fix: signDisplay plus sign behaviour fixed for NumberFormat

shubhamguptadream11 opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The two fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/hermes) and create your branch from `main`.
  2. If you've fixed a bug or added code that should be tested, add tests!
  3. Ensure it builds and the test suite passes. [tips](https://github.com/facebook/hermes/blob/HEAD/doc/BuildingAndRunning.md)
  4. Format your code with `.../hermes/utils/format.sh`
  5. If you haven't already, complete the CLA.
-->

## Summary

<!--
  Explain the **motivation** for making this change.
  What existing problem does the pull request solve?
-->
Fixes following issue: 
- https://github.com/facebook/hermes/issues/1466
- https://github.com/facebook/hermes/issues/1138
- https://github.com/facebook/hermes/issues/789

**Why this change is made?**

There are few cases where signDisplay is not being handled correctly.
Examples:
```
1.
new Intl.NumberFormat('de-DE', {
 style: 'currency',
 currency: 'EUR',
 signDisplay: 'exceptZero',
}).format(8537.71)

Output: +8,537,71+
Expected: +8.537,71 €

2.
new Intl.NumberFormat('ja-JP', {
 style: 'currency', 
 currency: 'JPY', 
 signDisplay: 'exceptZero' 
}).format(123456.789)

Output: +123,457
Expected: +￥123,457
```

**ChangeLog**

This PR updates the implementation and testing of the signDisplay functionality in the DecimalFormat class within the Hermes engine, specifically for Android API level 31 and above.

**Key Changes:**

Implementation:
- Integrated the **setSignAlwaysShown** method of DecimalFormat for API level 31 and above to control the display of the sign (+ or -) based on the signDisplay option. [For more detail about setSignAlwaysShown check [here](https://developer.android.com/reference/android/icu/text/DecimalFormat#setSignAlwaysShown(boolean))]
- For API levels below 31, maintained the existing logic for handling sign display, ensuring backward compatibility.

## Test Plan

<!--
  Demonstrate the code is solid.
  Example: The exact commands you ran and their output,
  screenshots / videos if the pull request changes the user interface.
-->

- Added a comprehensive set of test cases in **HermesIntlAndroidTest.java** to validate the behaviour of the signDisplay functionality, specifically for API level 31 and above.
- Test cases cover scenarios for signDisplay: `NEVER | ALWAYS | EXCEPTZERO`

shubhamguptadream11 commented:
@tmikov Whenever you have time, can you please check this PR? It solves the `signDisplay` issue reported twice on the platform.

neildhar commented:
Hey @shubhamguptadream11, thank you for looking into this and submitting this PR!

Could you describe the exact change here in more detail? It looks like there is more going on with this change, with regards to whether the positive/negative strings are checked and whether things are added to the suffix or prefix.

Alternatively, it looks like Android API level 31 adds [native support](https://developer.android.com/reference/android/icu/text/DecimalFormat#setSignAlwaysShown(boolean)) for this functionality. I think it would be preferable to add an API version check and use the newer API when possible, which should be available on most devices.

neildhar commented:
Looks great!

It looks like CI is failing, likely because the SDK version specified in our `build.gradle` is too low. Could you bump that, and add a test for the desired behaviour in `HermesIntlAndroidTest.java`?

shubhamguptadream11 commented:
Hi @neildhar,

I attempted to upgrade the `compileSdkVersion` to 31, but I encountered several issues due to deprecated APIs. For example, **InstrumentationTestCase** is deprecated as of API level 24 and needs to be replaced with **InstrumentationRegistry**. Given the scope of these changes, it would be more appropriate to handle this upgrade in a separate PR.

Could someone from the Hermes team take on the task of upgrading the compile/target version to the latest supported? or help me in upgrading that. Once that's done, I'll be able to raise my PR.

Thanks!

shubhamguptadream11 commented:
@neildhar Finally all the check are passed now.
Please review it once.

neildhar commented:
@shubhamguptadream11, thanks for bumping the SDK version and fixing up the deprecated APIs. Everything looks good, please add a test and update the summary, and I'll merge this.

shubhamguptadream11 commented:
@neildhar Added test cases and updated PR summary as well. Please check and let me know if anything else needed from my end.
Thanks again!

shubhamguptadream11 commented:
Hi @neildhar 

Just wanted to follow up on this PR. I’ve added the test cases and updated the summary as discussed. Could you please review it when you get a chance? Let me know if there’s anything else you need from my side.

Thanks again for your time!

neildhar commented on android/intltest/java/com/facebook/hermes/test/HermesIntlAndroidTest.java:
```diff
@@ -80,4 +85,46 @@ public void testDateTimeFormatCaseInsensitivity() {
+    if (Build.VERSION.SDK_INT >= 31) {
+      try (JSRuntime rt = JSRuntime.makeHermesRuntime()) {
+        rt.evaluateJavaScript(
+            "var nf = new Intl.NumberFormat('de-DE', { exceptZero: 'never', currency: 'EUR' });\n" +
```
This test seems odd, is it correct? I'm not aware of `exceptZero` being an option to `NumberFormat` (as opposed to `signDisplay: "exceptZero"`), and I don't believe the currency is printed unless you set `style: "currency"`.

If this is incorrect, we should understand why this test didn't fail.

shubhamguptadream11 replied:
I've corrected the test case to properly reflect the intended behaviour by using signDisplay: 'exceptZero' and including the style: 'currency' option.

shubhamguptadream11 commented:
Hi @neildhar ,

Could we please check and review this PR when you have a moment?

neildhar commented:
@shubhamguptadream11 Could you clarify if the test is actually running or if you ran it locally? It seems odd that the previous version of the test was working.

shubhamguptadream11 commented:
@neildhar 
I haven't found documentation specifically for running Android test cases like HermesIntlAndroidTest.java. I was under the assumption that these tests are automatically run on the CI after raising a PR. It's surprising to me as well that the previous tests passed without issues. Could you provide any guidance on how to properly run these tests locally or confirm if they are indeed executed on the CI?

shubhamguptadream11 commented:
@neildhar Please check the PR once. I had shared few screenshots above about test cases results.

shubhamguptadream11 commented:
Hi @neildhar 
I hope you're doing well. I wanted to follow up on the status of PR. If there's anything you'd like me to adjust, just let me know.

Thank you for your time and support!

shubhamguptadream11 commented:
Hi @neildhar, I hope you are doing well. Can you please take a moment to review this PR.
Thanks!
